src/S3/S3ClientTrait.php

```diff
@@ -273,6 +298,42 @@ public function doesObjectExist($bucket, $key, array $options = [])
         );
     }
 
+    /**
+     * @see S3ClientInterface::doesObjectExistV2()
+     */
+    public function doesObjectExistV2(
+        $bucket,
+        $key,
+        $includeDeleteMarkers = false,
+        array $options = []
+    )
+    {
+        $command = $this->getCommand('HeadObject', [
+                'Bucket' => $bucket,
+                'Key'    => $key
+            ] + $options
+        );
+
+        try {
+            $this->execute($command);
+            return true;
+        } catch (S3Exception $e) {
+            $response = $e->getResponse();
+
+            if ($includeDeleteMarkers
+                && (!empty($response)
```

Comment:
Maybe extract this into a hasDeleteMarker method that checks if a response exists and returns whether or not that header is there?  Condense a a bit of code.

Reply:
totally agree. just took a second glance and that looks pretty ugly